src/section/prefix_map.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright 2020 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::xor_space::{Prefix, XorName};
+use std::{
+    collections::{btree_map, BTreeMap},
+    fmt::{self, Debug, Formatter},
+    iter::FromIterator,
+};
+
+/// Map whose keys are `Prefix`es with the additional invariant that no two prefixes that are
+/// compatible (one is extension of the other) can be in the map at the same time.
+#[derive(Default, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub(crate) struct PrefixMap<T>(BTreeMap<Prefix<XorName>, T>);
+
+impl<T: Clone> PrefixMap<T> {
+    /// Create empty `PrefixMap`.
+    pub fn new() -> Self {
+        Self(Default::default())
+    }
+
+    /// Inserts new entry into the map. Replaces previous entry at `prefix` or any of its ancestors
```

Comment:
I wonder if it makes sense to only remove Prefix when we have both descendants here?  (i.e. after a split)

Reply:
Ah this is covered in a later commit :+1: